[MRG] support local genome collections (including private genomes) (#130)
dib-lab/genome-grist: doc/configuring.md

## Preparing information on private genomes

If you want to use unpublished or private genomes with genome-grist, you'll need to provide your own sourmash database, your own set of genome files and information, and your own taxonomy spreadsheet. Luckily this is all pretty straightforward and we provide tools to help you! Read on!

Note that you can absolutely combine Genbank with your own databases here, or just use your own databases. (If there are overlapping identifiers, the private genomes are chosen first; you might want to do this if you already have a bunch of the Genbank genomes downloaded already, for example.)

### Choosing identifiers for your genomes

You'll need to choose unique identifiers for your genomes. genome-grist requires that your identifier does not have a space, colon (:), or forward slash (/) in it; everything else should be fine.

### Preparing your genome files

You'll need one FASTA file per genome (gzip or bz2 compressed is fine). The filename doesn't matter. It's probably easiest if they're all in one directory, although this isn't necessary.

> We suggest naming at the first sequence in each FASTA file with the identifier at the start, space delimited - for example `MY_ID_1.1 first_sequence_name is very special` .

Reviewer: Will redundant sequence IDs affect the results? The entire header is unique but redundant IDs.

Author: We don't enforce anything about the IDs in sourmash, and the way genome-grist is coded, overlaps/non-unique IDs will end up resolving to just one genome (I don't recall if it's the first encountered, or what - I'll take a look).

I'll think if there's a good place to put checks in for non-unique IDs. Should be.

Author: https://github.com/dib-lab/genome-grist/issues/137

> We suggest naming at the first sequence in each FASTA file with the identifier at the start, space delimited - for example `MY_ID_1.1 first_sequence_name is very special` .

Reviewer: Can genome-grist or sourmash handle this renaming to minimize the effort for the user to pre-process the data especially for the software to run?

one suggestion is for the user to allow sourmash to inject unique identifiers and generate a mapping file after the processing. + replacing/trimming the special characters in the sequence header like the ":" or "/"

Author: I see two likely scenarios -

1. Someone has a few (or a few dozen) genomes, and they do the renaming on their own by hand.
2. Someone has a few thousand genomes, and then they know how to script.

If it turns out to a key need we can work on it, but for now I don't want to invent new problems for myself - every piece of software that we write demands support, after all!

> We suggest naming at the first sequence in each FASTA file with the identifier at the start, space delimited - for example `MY_ID_1.1 first_sequence_name is very special` .

Reviewer 2: This is a very confusing sentence for me...I don't really know what it's suggesting, so I can't make a suggesting for how to fix it 😄

Reviewer 2: Oh...now I see. using name from first. hm. I hate this suggestion. I dislike things that suggest altering canonical files. that feels like a recipe for mistakes. Could we suggest --name parameter instead of --name-from-first or something?

Author: will think on't.

Author: so, the problem is exactly what you see in the podar-ref example: large-scale sketch (using wildcards on genome files) really only works with `--name-from-first` at the moment. I don't see an easy resolution for this, but it's something we are thinking about over in sourmash so 🤷 . I'll update the sentence above a bit and then let this rest, because I don't intend to do anything about it in the near future...

Author: adjusted in 07786d45 and later to read:
>For now, we suggest naming the first sequence in each FASTA file with the genome identifier at the start, space delimited - for example `MY_ID_1.1 first_sequence_name is very special`. This will allow sourmash to name the resulting signature with the right identifier using `--name-from-first` (see below).

### Creating one or more sourmash databases

> You'll need to provide at least one sourmash database for your private collection to genome-grist under the config parameter `private_databases`, which takes a list of paths to sourmash database locations.

Reviewer 2: Can you provide multiple databases, one being your private collection and one being e.g. GTDB?

Author: yep. will work on making this clearer.

Actually, it makes me wonder if we even need to distinguish between private and genbank databases at this point in the config. Will revisit.

Reviewer: On farm for example, I can use the precomputed genbank as private genomes. So, I think *local genomes* will be more descriptive here.

Author: added to checklist to consider, and updated text to read:
>You can mix private databases with genbank databases, but 
you'll need to provide one or more sourmash databases for any private collections. You can do this with the config parameter `private_databases`, which takes a list of paths to sourmash database locations.

Author: ahh, missed @mr-eyes comment which hit between me typing mine and me hitting 'resolve' :).

I consider this one of those fine distinctions that experts will be able to figure out immediately, but will only confuse newbies. See also the text above,
Note that you can absolutely combine Genbank with your own databases here, or just use your own databases. (If there are overlapping identifiers, the private genomes are chosen first; you might want to do this if you already have a bunch of the Genbank genomes downloaded already, for example.)

I do recognize that there is potential confusion between local, private, unpublished, etc. and I've tried to write it somewhat clearly, but I'm done for now :)

Sketch all your genomes with the following command:
```
sourmash sketch dna -p k=31,scaled=1000
```

If you've named your genomes so that the first sequence contains the identifier, you can add `--name-from-first` and then the sequences will be named the right thing for the next step.

> If not, you'll need to manually adjust the names of the signatures produced by `sourmash sketch`. (You can do this with `sourmash sig rename`, but there's no simple way to do this in bulk.)

Reviewer 2: idk, the `--merge` parameter or `sourmash sig rename` feel like good ways...

Author: But those don't work in bulk, right? One... signature... at... a... time... just doesn't scale.

I think what we really need is something like `sourmash sketch fromfile` or a bulk renaming approach or something.

Author: err, [sourmash sketch fromfile](https://github.com/sourmash-bio/sourmash/issues/1671#issuecomment-987016199) ref.

Author: resolving for now.

> Once you have all your genome signatures, you can combine them into a single file with 

Reviewer 2: ```suggestion
Once you have all your genome signatures, you can create a sourmash database with
```

```
sourmash index output.sbt.zip *.sig
```

> and that will then be your sourmash database that you can cherish and treasure!

Reviewer 2: ```suggestion
and then you can cherish and treasure sourmash database forever!
```

> If you have lots of genomes (1000 or more?) there are other approaches that might make your life more convenient; just ask for suggestions on [the sourmash issue tracker](https://github.com/dib-lab/sourmash/issues).

Reviewer 2: none come to my mind immediately, could this be less cryptic so I wouldn't necessarily have to open an issue just to satisfy my curiosity?

Author: 😆 fine

Author: added:
>If you have lots of genomes (1000 or more?) we suggest using a workflow system 
to sketch and rename your genomes appropriately; please ask us for some examples
 over on [the sourmash issue tracker](https://github.com/dib-lab/sourmash/issues).

in 07786d45

> We chose k=31 above (in the `sourmash sketch` command) because that matches our default parameters, and we have provided Genbank databases for k=31 (as well as k=21 and k=51). But the only real requirement here is that all your databases support the requested k-mer and scaled sizes.

Reviewer: So there are 3 scenarios here.

1. Use genome-grist to query GenBank or any precomputer DB, in this case, the user will stick to k(21,31,51) and the supported scales.
2. Use genome-grist to query precomputed DB + private genomes. This will be typical of the first case.
3. Use genome-grist to query only private genomes, in this case, the user is free to select the k-size and scale

Author: Yep!

> We chose k=31 above (in the `sourmash sketch` command) because that matches our default parameters, and we have provided Genbank databases for k=31 (as well as k=21 and k=51). But the only real requirement here is that all your databases support the requested k-mer and scaled sizes.

Reviewer 2: ```suggestion
We chose k=31 above (in the `sourmash sketch` command) because that matches our default parameters, and we have provided Genbank and GTDB databases for k=31 (as well as k=21 and k=51). But the only real requirement here is that all your databases support the same requested k-mer and scaled sizes.
```

### Providing your genomes to genome-grist

> You'll also need to provide your genome files to genome-grist, along with their "display name", which will be used in reporting.  The information will be provided via the config parameter `private_databases_info`, which takes a list of paths to info file CSVs

Reviewer 2: ```suggestion
You'll also need to provide your genome files to genome-grist, along with their "display name", which will be used in reporting.  The information will be provided via the config parameter `private_databases_info`, which takes a list of paths to info file CSVs.
```

> **First**, genome-grist needs the genomes in their own individual files, in one or more directories. The files need to be named by their identifiers in the format `{ident}_genomic.fna.gz`, and must come with an "info file" that contains their identifier, a display name, and the location of the genome file (which _must_ be named as above).

Reviewer 2: can you show a `head` of a file like this? that would be super helpful at this point

Author: added in 07786d45787

> genome-grist has a utility to help set this all up! The script `genome_grist.copy_private_genomes` will take in a list of FASTA files containing genome(s), read the header of the first sequence to find the identifier for that genome, and then copy it into a directory for you. (see "Step 3", below, for execution instructions for this script). It will also output a provisional info file, which you can edit.

Reviewer: Will that cause data redundancy? I'd suggest symlinking them in the new directory with new names.

Author: Yes, it will. Unfortunately, Snakemake doesn't like symlinks. So it'd have to be hard links. I'll put this down under future options...

But, it's a minor problem for private collections that are under a few hundred genomes; microbial genomes are small and cute! I would have more of a problem duplicating the metagenome samples themselves, because they can be QUITE large.

Reviewer: I got your point, thanks :)

For the Snakemake symlinks problem, maybe they've resolved this? because all the times I used genome-grist my samples were symlinks.

> **Second**, for each genome, genome-grist also needs a separate `{ident}.info.csv` file, containing just the identifier and the display name. This needs to be in the same directory as the genome itself.

Reviewer 2: again, a head could be useful. Something like "the output of the utility script `genome_grist.make_info_file` should look something like:

Author: added in 07786d45787
